cosmos_rl/policy/model/hf_models/patch.py

```diff
@@ -93,3 +94,222 @@ def patch_forward(self, **kwargs) -> torch.LongTensor:
+    offset = 0
+    for length in lengths:
+        mask[offset : offset + length, offset : offset + length] = torch.tril(
+            torch.ones((length, length), dtype=torch.bool, device=device)
+        )
+        offset += length
+    return mask
+
+
+# Get packed sliding attention mask
+def get_packed_sliding_attention_mask(lengths, sliding_window, device):
+    # lengths: list of sequence lengths
+    L = sum(lengths)
+    mask = torch.zeros((1, 1, L, L), dtype=torch.bool, device=device)
+    cur = 0
+
+    for L in lengths:
+        start = cur
+        end = cur + L
+        for i in range(start, end):
+            valid_start = max(start, i - sliding_window + 1)
+            valid_end = i + 1
+            mask[0, 0, i, valid_start:valid_end] = True
+        cur = end
+    return mask
+
+
+def make_new_self_attn_forward(original_attn_forward):
+    def self_attn_forward(self, hidden_states, *args, **kwargs):
+        valid_input_len = kwargs.get("valid_input_len", None)
+        if valid_input_len is not None:
+            if hasattr(self, "is_sliding") and self.is_sliding:
+                assert (
+                    hasattr(self, "sliding_window") and self.sliding_window is not None
+                ), "sliding_window must be set for sliding attention"
+                attention_mask = get_packed_sliding_attention_mask(
+                    valid_input_len.tolist(), self.sliding_window, hidden_states.device
+                )
+            else:
+                attention_mask = get_packed_attention_mask(
+                    valid_input_len.tolist(), hidden_states.device
+                )
+            kwargs["attention_mask"] = attention_mask
+        return original_attn_forward(hidden_states, *args, **kwargs)
+
+    return self_attn_forward
+
+
+def sequence_packing_forward_patch(hf_config: AutoConfig, hfmodel):
+    patch_success = False
+    try:
+        if hf_config.model_type in SEQUENCE_PACKING_FORWARD_PATCH_FUNCTIONS:
+            SEQUENCE_PACKING_FORWARD_PATCH_FUNCTIONS[hf_config.model_type](
+                hfmodel.model
+            )
+            patch_success = True
+        else:
+            if not hfmodel.is_vlm:
+                SEQUENCE_PACKING_FORWARD_PATCH_FUNCTIONS["llm"](hfmodel.model)
+                patch_success = True
+            else:
+                logger.warning(
+                    f"Failed to patch sequence packing forward for {hf_config.model_type}, supported models: {SEQUENCE_PACKING_FORWARD_PATCH_FUNCTIONS.keys()}"
+                )
+    except Exception as e:
+        logger.error(f"Failed to patch sequence packing forward: {e}")
+    return patch_success
+
+
+def sequence_packing_forward_qwen3_vl_patch(model):
+    original_forward = model.language_model.forward
+
+    def sequence_packing_forward_qwen3_vl_inner(*args, **kwargs):
+        valid_input_len = kwargs.get("valid_input_len", None)
+        if valid_input_len is not None:
```

Comment:
Same issue here. The position id and mask related operations seems to be executed for each layer, which might be duplicated?